### PR TITLE
livereload now works with grunt-contrib-watch #27

### PR DESCRIPTION
**Pull request by thanpolas**
- livereload now works with grunt-contrib-watch
- readme updates

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -108,17 +99,17 @@ module.exports = function (grunt) {
...
 
-  grunt.registerTask('default', ['livereload-start', 'connect', 'regarde']);
+  grunt.registerTask('default', ['livereload-start', 'connect', 'watch']);
 };
 ```
 
 
 ## Release History
-
+ * 2013-03-23   v0.2.0   Added support for *grunt-contrib-reload*.
```
Comment (sindresorhus) [inline README.md]: grunt-contrib-watch?

File: tasks/livereload.js [hunk 2]
```diff
@@ -42,5 +43,16 @@ module.exports = function (grunt) {
   grunt.registerTask('livereload-start', 'Setup livereload to alert your browser when a file has changed', function () {
     // Start a websocket server in the background
     server = utils.startLRServer(grunt, this.async());
+
+    // listen for watch events
+    grunt.event.on('watch', function(action, data){
+      if ('changed' !== action) {
+        return;
+      }
+      var files = Array.isArray(data) ? data : [data];
```
Comment (sindresorhus) [inline tasks/livereload.js]: Is this really needed? grunt-contrib-watch docs says it should return _a_ filepath

File: README.md [hunk 3]
```diff
@@ -108,17 +99,17 @@ module.exports = function (grunt) {
 
   });
 
-  grunt.loadNpmTasks('grunt-regarde');
+  grunt.loadNpmTasks('grunt-contrib-reload');
```
Comment (sindresorhus) [inline README.md]: ?

File: tasks/livereload.js (hunk 2 above)
Comment (thanpolas) [inline tasks/livereload.js]: `data` is passed as a string, `server.changed()` requires an array...

**Comment by thanpolas** [comment]
@sindresorhus, @shama the problem with this implementation is that _livereload_ gets triggered for **all** watch targets.

consider this:

``` js
    watch: {
      less: {
        files: ['_styles/**/*.less'],
        tasks: 'lessCopy'
      },
      staticSources: {
        files: ['_site/**'],
        tasks: ['livereload']
      } 
   }
```

If change a `.less` file is changed, a `watch` event will be dispatched and reload will trigger... and then when lessCopy finishes and updates the css in `_site` livereload will run again...

This can be annoying all the way to livereload server crashing when many files are involved...

ideas?

**Comment by shama** [comment]
@thanpolas That is the desired behavior though, correct? For most, the amount of simultaneously changed files should be one to a few. Do you know at what amount of files the livereload server will start crashing?

EDIT: Oh also, I'll update the watch to not require the `tasks` property.

**Comment by sindresorhus** [comment]
@thanpolas I'm not sure how we can prevent that. People should only add watch targets for compiled assets, otherwise it will reload twice, as you said, and do a full page refresh on the first, since it can't match the .less files with anything.

File: tasks/livereload.js (hunk 2 above)
Comment (sindresorhus) [inline tasks/livereload.js]: Yes, so why not use `[data]` directly?

**Comment by sindresorhus** [comment]
> Backwards compatibility

We're still in pre1.0 so I don't see at as a problem to do a breaking change on a `minor` release. People have specified `~x.x.x` which means it won't break unless they upgrade manually. We'll make sure to make it clear in the changelog. @shama wdyt?

I will `npm deprecate` the regarde task when this lands.

**Comment by thanpolas** [comment]
>  People should only add watch targets for compiled assets

You can't say that... What happens to all me pre-processing files? Who will watch them?

I don't think it would be a stretch to merge livereload with watch @shama...

``` js
    watch: {
      less: {
        files: ['_styles/**/*.less'],
        tasks: 'lessCopy'
      },
      staticSources: {
        livereload: true, // <----------- just one option
        files: ['_site/**'],
        tasks: ['livereload']
      } 
   }
```

**Comment by shama** [comment]
@thanpolas I'd rather not implement live reload into the watch task. That way users can use other watch tasks (or some other triggering mechanism) with live reload. Let's just focus on implementing this and making the contrib tasks interoperable for now.

@sindresorhus Sounds good to me. Thank you for taking care of all this. I really appreciate it.

**Comment by thanpolas** [comment]
roger @shama.

So where are we now? Is this good to go? Need me to do something else? @sindresorhus

File: tasks/livereload.js (hunk 2 above)
Comment (shama) [inline tasks/livereload.js]: `data` should always be a string. Naming it `filepath` might be better too.

**Comment by shama** [comment]
Another thought; the `watch` event is fired as soon as a file is modified and not when the watch is actually triggered (the watch is debounced, the event is not). Would it be a good idea to have the `on('watch')` just collected modified files then run a debounced function to trigger `server.changed`?

**Comment by thanpolas** [comment]
@shama done

**Comment by shama** [comment]
Thanks Thanasis. Although I think it would be better to just use the lodash debounce. Like this:

``` js
var changedFiles = Object.create(null);
var onChanged = grunt.util._.debounce(function() {
  server.changed({body:{files:Object.keys(changedFiles)}});
}, 250);
grunt.event.on('watch', function(action, filepath) {
  changedFiles[filepath] = action;
  onChanged();
});
```

Also as @sindresorhus said, since we're doing a pre 1.0.0 minor release, we can break backwards compatibility. So why not remove the `livereload-start` task and just have the `livereload` task?

**Comment by thanpolas** [comment]
What does `_.debounce` offer over `setTimeout`?

The `livereload-start` task is expected to be invoked once and only once to setup the event listener.

The `livereload` task on the other hand is expected to get invoked as an action for a watch task. So if we add the event listener in `livereload` we break backwards compatibility.

Finally, do we not care for the `action` to explicitly be `changed` ?

**Comment by shama** [comment]
Using `_.debounce` instead just means less code here. Also if we do our own debounce here then the question rises why not in the other contrib plugins... then it becomes a thing. So since we have access to it now in Grunt v0.4, easier just to use it. Down the road I imagine if/when lodash isn't packaged with Grunt and only using 1 lodash function; it might make more sense then. For now, let's just use it.

I don't think we should worry about breaking backwards compatibility, as long as we do semver right and update the changelog. I think it would make this task simpler just having the one `livereload` task since we're using the event now. We would only need to do `grunt livereload watch` and don't need to instruct the watch task to run any task to trigger the livereload. Just as long as the task/server is running.

The other option is to not use the event and keep the two tasks. Then have the watch trigger livereload through a task. The event way just seems the simplest to me.

I think it should trigger on all `action`s. If I add or delete a file, I would want to see the effect just as much as changing a file.

**Comment by thanpolas** [comment]
> The other option is to not use the event and keep the two tasks.

@shama if you can do that on watch'es end then it would be the best solution. Event triggering for all watched files will come back and bite us. I don't want livereload to trigger when i update a .less file, using events livereload will trigger with any watch event.

**Comment by sindresorhus** [comment]
> You can't say that... What happens to all me pre-processing files? Who will watch them?

Should have been more specific. What I meant is that users should only _trigger_ livereload on compiled assets like .css/.js/.png etc and not .less. So instead of a loose globbing like `files: ['_site/**']` they should use `files: ['_site/.{js,css,jpg,png}']`

**Comment by thanpolas** [comment]
Yep, we are talking about the same thing @sindresorhus 

So, if we use the `watch event` livereload will trigger with any triggered watch event no matter the globbing pattern / tasks.

@shama what can the watch task do for this?

**Comment by sindresorhus** [comment]
@thanpolas yeah, I just realized when rereading the thread. I would prefer the event method, since having 2 tasks is suboptimal, but so is not being able to specify what files to watch for livereload.

**Comment by shama** [comment]
@thanpolas Actually you raise a good point. If we only have the one task, then the user can't configure which watch target will do the reload.

So maybe having `livereload-start` setup the server and start the event listener to collect changed files. Then have the watch trigger the `livereload` task feeding the changed files to `server.changed`. This would eliminate the need for a `debounce` as well.

**Comment by thanpolas** [comment]
> Then have the watch trigger the livereload task feeding the changed files to server.changed.

Yes, how do we do this @shama ?

**Comment by shama** [comment]
``` js
var server;
var changedFiles = Object.create(null);
grunt.registerTask('livereload-start', function() {
  server = utils.startLRServer(grunt, this.async());
  grunt.event.on('watch', functon(action, filepath) {
    changedFiles[filepath] = action;
  });
});
grunt.registerTask('livereload', function() {
  if (!server) { grunt.fatal('Please run the livereload-start task first.'); }
  server.changed({body:{files:Object.keys(changedFiles)}});
  changedFiles = Object.create(null);
});
```

**Comment by thanpolas** [comment]
two things here:
1. When watch triggers the `livereload` task it seems to be on a spawned process? Because the `server` var is undefined... I have no options in the watch configuration...
2. The `changedFiles` Object will still contain every file that triggered the watch event (files.less, .png, .jade, etc)

**Comment by shama** [comment]
1. Yes, with this approach the watch `nospawn` option must be set.
2. `changedFiles` will only contain the changed files; which in most cases will only be 1 to a few. In the cases where `tiny-lr` crashes from sending too many files to it... I say that is an issue that should be investigated on `tiny-lr` (also I haven't used tiny-lr with triggering a large amount of simultaneous files but I would expect it shouldn't care). But in any case, I would expect the user to more explicitly target their files as @sindresorhus said.

So I'm on the fence about each of these two methods. One is simpler (just events, tasks can spawn if the user wants) but the other is more configurable (two tasks, must nospawn but livereload can be configured to only trigger per watch target). Or maybe a better method we're missing? @sindresorhus?

**Comment by sindresorhus** [comment]
Hmm, some other ideas:
1. Watch tasks adds ability to subscribe to sub-events: `.on('watch:target')`, and only get the files from that target. You would then specify these wanted watch targets in the livereload config object.
2. Watch event embeds the target which let's the livereload task get the target files array and compare. You would then specify these wanted watch targets in the livereload config object.
3. I'm beginning to think it would be easier to just use the watch as a lib and have the user specify the files directly in the livereload config.

**Comment by shama** [comment]
All doable. Someone else was asking about passing the target in the watch event as well.

FWIW, while implementing gruntjs/grunt-contrib-watch#32, the `taskrun` lib there has really become it's own module (currently unpublished, working on tests). I was talking to @cowboy about whether it should become a standalone module or stay within the watch task. Also I suppose we could just expose it and other tasks could use the watch by `require('grunt-contrib-watch')`. I think the only reason it should stay within `grunt-contrib-watch` is because it still heavily relies on a grunt task, so being standalone might confuse people attempting to use it as just a module.

**Comment by thanpolas** [comment]
mixing both your ideas we can do this implementation without requiring the watch task to fail:

``` js
var server;
var changedFiles = Object.create(null);
grunt.registerTask('livereload-start', function() {
  server = utils.startLRServer(grunt, this.async());
  grunt.event.on('watch', functon(action, filepath) {
    changedFiles[filepath] = action;
  });
});
grunt.registerTask('livereload', function() {

  if (!server) { grunt.fatal('Please run the livereload-start task first.'); }

  var files = grunt.data.files;
  var canReloadFiles = grunt.file.expand(files);
  var reloadFiles = changedFiles.filter(function(file){ return  canReloadFiles.indexOf(file) >= 0;});

  server.changed({body:{files:Object.keys(changedFiles)}});
  changedFiles = Object.create(null);
});
```

### Gruntfile Config

``` js
    livereload: {
      buildfiles: {
        files: ['_site/**']
      }
    },
    watch: {
      less: {
        files: ['_styles/**/*.less'],
        tasks: 'lessCopy'
      },
      staticSources: {
        files: ['_site/**'],
        tasks: ['livereload']
      } 
   }
```

My only concern is the `nospawn` option

**Comment by shama** [comment]
I'm convinced. Number 3, use the watch as a lib, seems to satisfy all the concerns raised. Then this task wouldn't require the user to connect up tasks in an awkward way. This task becomes self sufficient and can handle config however it wants. It would operate just like all the other tasks.

I'll finish up gruntjs/grunt-contrib-watch#32 and then work on exposing the watch as a lib. Then we can implement here. Sound good?

**Comment by thanpolas** [comment]
So if we go with 3, that means that the livereload task will launch it's own watch spawned process?

**Comment by shama** [comment]
The actual watching part is handled by `gaze` and I don't think that part of the watch is worth wrapping. `gaze` can be consumed as is and will always continue to work with Grunt. I think the task running part should be exposed in an API something like this:

``` js
var taskrun = require('grunt-contrib-watch');
// or maybe make this: var taskrun = require('grunt-lib-taskrun');

var Gaze = require('gaze').Gaze;
var server;

grunt.registerTask('livereload', function() {
  var options = this.options({port:35729});
  var done = this.async();
  if (!server) { server = utils.startLRServer(grunt, done); }

  // initialize taskrun and get normalized targets from config
  var targets = taskrun.init(grunt.config(this.name || 'livereload'), options, done);
  var changedFiles = Object.create(null);

  taskrun.on('done', function() {
    server.changed({body:{files:Object.keys(changedFiles)}});
    changedFiles = Object.create(null);
  });

  function onChange(action, filepath) {
    changedFiles[filepath] = action;

    // queue targets to run
    taskrun.queue(target);

    // run the queue (debounced)
    taskrun.run();
  }

  Object.keys(targets).forEach(function(target) {
    new Gaze(target.files, target.options, function(err) {
      this.on('all', onChange);
    });
  });
});
```

Which is basically the watch task with the `server.changed` and brings us back to Thanasis's idea: add live reload to the watch task. Although I still think we should keep them separate as I believe this task and the watch task will continue to individually grow.

The user can choose whether to spawn tasks or not by setting the `nospawn` option per target (or across the task) just like the other plugins.

**Comment by sindresorhus** [comment]
> I'll finish up gruntjs/grunt-contrib-watch#32 and then work on exposing the watch as a lib. Then we can implement here. Sound good?

Yes :)

> Although I still think we should keep them separate as I believe this task and the watch task will continue to individually grow.

I agree.

## 

@shama @thanpolas can we realistically get this done by the 28th? Anything I can help out with?

**Comment by shama** [comment]
@sindresorhus I don't think it will be done by the 28th. I'll have the changes made to the watch task by then but I still want to run the code by @cowboy and @tkellen first. So maybe they can give us an extension as progress is being made and I think we've got a good solution hashed out here.

**Comment by tkellen** [comment]
As long as there is active communication and actual progress being made (with an end in sight), take the time you need!  My heavy-handedness with the deadline was because there was no communication at all for several weeks.

**Comment by sindresorhus** [comment]
Ok, sounds good.

**Comment by thanpolas** [comment]
@sindresorhus @shama  I'm not very confident i understand what needs to be done...

If livereload will consume its own stack of Gaze then it would essentially be another watch task. And we know _there can be only one_, unless parallel execution is in order so we can run livereload and watch simultaneously...

Please give me bullets / specs of what i need to do or let me know if someone else will eventually do it...

**Comment by shama** [comment]
@thanpolas It's on me atm. I'll ping you if you would still like to implement this once I finish up the work on the watch task. With this solution the watch task will be a dependency of this task. So it is another watch task but I see it as modified version of the grunt-contrib-watch task.

Maybe it makes sense to merge them later, I'm not sure. It makes sense to me now to implement the parts both of these tasks have in common into the watch task and have this task extend from that. Since doing so will allow this task to become standalone and easier for the end user to config.

**Comment by thanpolas** [comment]
@shama will the user be able to run the watch and the livereload task simultaneously?

**Comment by shama** [comment]
@thanpolas Why would they need to?

**Comment by thanpolas** [comment]
``` js
    watch: {
      less: {
        files: ['_styles/**/*.less'],
        tasks: 'lessCopy'
      },
      staticSources: {
        files: ['_site/**'],
        tasks: ['livereload']
      } 
   }
```

I want a watch for all my pre-processing tasks (less, sass, jade, whatnot) ... they trigger, run, generate a .css, then i want livereload to trigger.

If livereload extends watch, then this flow will not be possible. Not happy with that.

**Comment by shama** [comment]
@thanpolas Hmm well let me run all this by @cowboy and get his opinion.

**Comment by shama** [comment]
Maybe this, with livereload extending the watch, simultaneous use might work:

``` js
watch: {
  less: {
    files: ['_styles/**/*.less'],
    tasks: 'lessCopy'
  },
},
livereload: {
  staticSources: {
    files: ['_site/**'],
    // If no tasks are specified, just trigger the livereload
  },
},
```

Although you can't run both with `grunt watch livereload` as each will keep alive. I can only think of hacky ways to get around that atm. So needs more discussion. I'll ping @cowboy soon.

**Comment by thanpolas** [comment]
@shama yes. That's the proposal i did [in this comment a tid bit above](https://github.com/gruntjs/grunt-contrib-livereload/pull/32#issuecomment-15367958).

If we have a green light with that, there is no need for watch to do anything and livereload will be ready and working in less than a day.

**Comment by jwebcat** [comment]
@thanpolas @sindresorhus @shama  I would also LOVE this feature as I am hoping to do very close to the same thing, triggering Live Reload after the processing / generating of the site/ files is done.

as of now All methods I have tried working with jekyll and Live Reload have left me feeling a bit vexed. Using Grunt has been the smoothest flow yet. I hope this feature is implemented soon.

Thanks for all your hard work on this :+1:

**Comment by itsjamie** [comment]
It seems to make sense to me for livereload to just be another task that watch can call on certain types of files.

This would allow a similar flow to what is currently being used the wild with something like..

``` javascript
watch: {
  js: {
    files: ['**/*.js', '!node_modules/**/*.js'],
    tasks: ['jshint', 'concat', 'uglify', 'livereload']
  }
}
```

This would be the "holy grail" solution in my eyes.

**Comment by shama** [comment]
Ok, let's throw that idea out then. I would really like to have livereload work with spawning. At my work we run the watch on our dev server for months. Even with duck punching `grunt.fail`, a nospawn watch task doesn't make it through a single busy day.

So in order to use both tasks **simultaneously**, with **spawn and nospawn**, **configurable per target** and overall just **easy to configure**... it seems we have two options:
1. **Add a `livereload` option to the watch task and turn this task into a lib**<br/>
   If the option is set, the watch will trigger the reload after the task runs:

``` js
    watch: {
      one: {
        files: ['src/*.js'],
        tasks: ['concat'],
        options: { livereload: true },
      }
    }
```
1. **Have the livereload task ping the reload server**<br/>
   With a task that maybe looks something like this (untested):

``` js
    module.exports = function(grunt) {
      var request = require('request');

      var server;
      var changedFiles = Object.create(null);
      grunt.event.on('watch', function(action, filepath) {
        changedFiles[filepath] = action;
      });

      grunt.registerTask('livereload', function() {
        var done = this.async();
        var options = this.options({port: 35729});

        if (!server) {
          // this url needs to be determined better
          var triggerUrl = 'http://localhost:' + options.port + '/changed';

          request.post({
            url: triggerUrl,
            json: { files: Object.keys(changedFiles) }
          }, function (error, res) {
            if (!error && res.statusCode === 200) {

              // Triggered by a spawning watch task
              changedFiles = Object.create(null);

            } else {

              // This is the first run, start the server and trigger reload
              server = utils.startLRServer(grunt, done);
              server.changed({body:{files:Object.keys(changedFiles)}});
              changedFiles = Object.create(null);

            }
            done();
          });

        } else {

          // Triggered by a nospawn watch task
          server.changed({body:{files:Object.keys(changedFiles)}});
          changedFiles = Object.create(null);

        }
      });
    };
```

This way those who want `nospawn` can continue to have a fast response (with slightly slower only on the first run) and those who want spawn and don't mind the delay can still use this task.

**Comment by thanpolas** [comment]
:+1: for **1**, livereload is very thin and we'll save a lot of pain with making it an option to the watch task.

**Comment by joeybaker** [comment]
:+1: for **1** as well. It's always seemed kinda like a no-brainer to me that livereload and watch should be combined.

**Comment by shama** [comment]
Chatted with @cowboy about this today. He seems to like option 1 as well:

> it seems like the easiest way to do this would be to hard-code tiny-lr into watch and add it as an option and have no extra tasks

@sindresorhus What is your opinion on this? Am I missing any better options?

**Comment by jperl** [comment]
:+1:, commented to watch this

**Comment by joeybaker** [comment]
@jperl I'm gonna be "that" guy: :]
Screen Shot 2013-04-02 at 3 05 32 PM

**Comment by jperl** [comment]
@joeybaker thank you!

**Comment by zacparker** [comment]
:+1: Option 1

**Comment by sindresorhus** [comment]
Looks like option 1 is the way to go, though I'm not super happy about it either.

## 

_Bringing in @passy @kevva @sleeper in case you have any additional thoughts._

**Comment by cowboy** [comment]
@sindresorhus the way I see it, the watch task could completely contain tiny-lr. That way, it can start the tiny-lr server if necessary, tell it what files have changed, and then watch doesn't need to communicate with anything external to it.

Maybe tiny-lr can be integrated as a kind of middleware, so that a user could change its behavior or replace / augment it with something else?

The way I see it, tiny-lr is such a simple lib that it can be both implemented in a custom task (if someone needs it) AND can just be included in watch, since so many people want watch + lr functionality.

**Comment by sindresorhus** [comment]
> Maybe tiny-lr can be integrated as a kind of middleware, so that a user could change its behavior or replace / augment it with something else?

Not sure what that would be though. I would rather us extend later after actual need.

> The way I see it, tiny-lr is such a simple lib that it can be both implemented in a custom task (if someone needs it) AND can just be included in watch, since so many people want watch + lr functionality.

Makes sense. Didn't mean to say it was a bad solution. Only that in a perfect world I would have preferred them as separate tasks. Though this seems like a good solution.

**Comment by shama** [comment]
FWIW, I was against option 1. Gauging by the current open issues; merging with the watch means more support issues for me and more code to maintain in an already rapidly growing watch task.

Making livereload it's own watch task seemed like the ideal solution until @thanpolas pointed out users would not be able to use the tasks simultaneously. I've turned around on option 1 because it seems just like the ideal solution, turning livereload into a watch task, but with the ability to use livereload and plain watching simultaneously.

The watch task is going through a big refactor atm that will close most of the current open issues and was gearing to turn the watch task into a lib. See the `reload` branch, not livereload... named reload because it originally was implementing gruntfile reloading. That branch is almost ready to merge into master.

So after that is merged, I'll start implementing livereload into the watch under a `livereload` branch, unless anyone objects?

**Comment by tkellen** [comment]
No objections here.  Thank you for all the work you do Kyle!

**Comment by sindresorhus** [comment]
:+1:

**Comment by Snugug** [comment]
:+1: for Option 1 as well.

**Comment by Snugug** [comment]
What's the status on this? Really am in need of this for a project.

**Comment by shama** [comment]
I'm going to include it in the next release of the watch task. Should be ready by the end of this week.

**Comment by Snugug** [comment]
Awesome. Very excited

**Comment by cowboy** [comment]
We're ALL very excited.

**Comment by thanpolas** [comment]
:100:

**Comment by dgieselaar** [comment]
Awesome. Last piece of the puzzle.

**Comment by shama** [comment]
Live reload has been implemented on the [reload branch](https://github.com/gruntjs/grunt-contrib-watch/tree/reload) of the watch task. Please test if you've got a moment.
[Docs for using livereload are here.](https://github.com/gruntjs/grunt-contrib-watch/blob/reload/docs/watch-examples.md#live-reloading)

---

One part it does not address is supplying a connect middleware to inject the script tag into the HTML. I think how the `livereload.js` script gets included in the HTML should be up to user and the templating mechanism they've chosen. Currently the recommended options are:
1. Add the `http://localhost:35729/livereload.js` script tag yourself.
2. Or use the live reload browser extension.

I've noticed that the connect middleware is the source of many issues on this repo. I think if we are going to supply a connect middleware, it should be another lib. Such as `grunt-lib-livereload`. The main reason I don't think we should recommend this route is it would be extremely verbose to configure with connect, at least IMO:

``` js
grunt.initConfig({
  watch: {
    css: {
      files: ['public/scss/*.scss'],
      tasks: ['compass'],
      options: {
        livereload: true,
      },
    },
  },
  connect: {
    server: {
      options: {
        middleware: function(connect, options) {
          return [
            // Serve static files.
            connect.static(options.base),

            // Make empty directories browsable.
            connect.directory(options.base),

            // Inject the live reload script into the page
            require('grunt-lib-livereload')({
              port: 35729,
            }),
          ];
        },
      },
    },
  },
});
```

As opposed to just adding `<script src="http://localhost:35729/livereload.js"></script>` to your page, template layout or whatever mechanism you choose.

So I'm going to leave it out unless anyone objects. Thanks!

/cc @cowboy @tkellen @sindresorhus

**Comment by cowboy** [comment]
It seems much easier to just utilize a `development` flag to enable including some html like `<script src="http://localhost:35729/livereload.js"></script>` in the HTML page via templates, instead of using a complex middleware to get the same result. Then LR will work no matter what server you're using.

And it's much simpler.

**Comment by joeybaker** [comment]
FWIW, I never found the LR middleware to be reliable. I've always just used a `script` tag.

@shama – seems to work in my first tests. Thank you!

**Comment by Snugug** [comment]
EDIT!
NEVERMIND! Was on the wrong branch, this works for me! It's awesome!

**Comment by shama** [comment]
`grunt-contrib-watch@0.4.0` has been published. This issue is officially closed. \o/

Thank you everyone who participated and especially @thanpolas and @sindresorhus!

@Snugug Give the latest `v0.4.0` a try and if you run into an issue please open the issue on [grunt-contrib-watch](https://github.com/gruntjs/grunt-contrib-watch/issues).

Thanks!

**Comment by dgieselaar** [comment]
_applause_

**Comment by ericduran** [comment]
:clap:

**Comment by Snugug** [comment]
:clap:

**Comment by sindresorhus** [comment]
@shama fantastic! \o/

I will deprecate grunt-contrib-livereload and grunt-regarde and update their readmes as soon as we've landed the [change](https://github.com/yeoman/generator-webapp/pull/59) in yeoman.

**Comment by sethburtonhall** [comment]
Can this support a Live Reload time delay?

**Comment by passy** [comment]
@middle8media 

# [deprecated]

**Comment by sethburtonhall** [comment]
got it thx.

**Pull request closed without merging**